### PR TITLE
Add minimal initial API spec document

### PR DESCRIPTION
**Pull request by ImJasonH**
First part of [TEP-0012](https://github.com/tektoncd/community/pull/172)

[Rendered](https://github.com/imjasonh/pipeline/blob/api-spec/docs/api-spec.md)

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [n/a] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [only] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [yes] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [y] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

```release-note
Add initial minimal API specification document
```

/assign @bobcatfish 
/assign @vdemeester

### CODE REVIEW

File: docs/api-spec.md
Comment (dlorenc) [inline docs/api-spec.md:54]: Do we expect to have any expected lifetime or mandatory wait period as features move through here? I think wed' want to give platform implementors some time to add support for things so they don't fall out of compliance.

Comment (ImJasonH) [reply docs/api-spec.md:54]: I think in practice we'll end up having one, maybe even enforcing and documenting it, but it seems unnecessary at this early stage.

When we reach the point of first making an OPTIONAL surface REQUIRED we'll reach out to platform implementors and get their feedback about how soon they can support it, and work out a timeline. Over time I expect we'll find it normally takes N weeks/months/releases to achieve this, and we can document that as the default upgrade timeline.

File: docs/api-spec.md
Comment (bobcatfish) [inline docs/api-spec.md:151]: should we list REQUIRED, etc for the fields above as well? (e.g. i have no idea what severity is about, seems less important)

File: docs/api-spec.md
Comment (bobcatfish) [inline docs/api-spec.md:260]: nice, thanks for whittling this down! 

is StepState similar (i.e. based on pod status) or is this something we created ourselves?

File: docs/api-spec.md
Comment (bobcatfish) [inline docs/api-spec.md:252]: am i crazy or is it possible to have a step without a name

File: docs/api-spec.md
Comment (bobcatfish) [inline docs/api-spec.md:271]: []TaskRunResult?

Comment (bobcatfish) [reply docs/api-spec.md:271]: ( dont let this fool you into thinking i was this thorough with everything i looked at XD )

File: docs/api-spec.md
Comment (bobcatfish) [inline docs/api-spec.md:307]: do we want to say anything more about this? in order to support tasks in a pipeline properly, the system would need to support _something_ besides emptyDir

**Review by bobcatfish — COMMENTED** [review]
metaquestion: what do you see as the process to add more fields here going forward? e.g. how to add when expressions into here. "Modifying This Specification" starts to talk about it but i wonder if it'll make sense to create a process around this to guide contributors

overall looks good, just lots of minor feedback from me!

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:37]: I don't mind it being in a separate document. It's not currently phrased in terms of "platform implementations must satisfy this contract" so much as "users should expect this contract", but maybe that's just fine since the vast majority of readers will be users, and very few will be prospective platform providers.

In the meantime it's probably enough to just say "this isn't the runtime contract, that's over there" and when/if we get any confusion we can resolve it. Launch and iterate 😄

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:45]: Oh probably just because the original draft was written when 1.13 was the minimum required. 👴 

AFAIK no relevant fields have been removed/changed since 1.13, so it's technically still true. 🤷

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:64]: Good question, one I should answer better in the doc 😄 

Basically, in the scope of the API spec, it's an arbitrary string that represents a container of taskruns. In the k8s implementation it'll be a namespace, but we should be meaningfully vague here to allow other platforms to collect resources however they want, including enabling hierarchies (e.g., something like `/apis/tekton.dev/v1beta1/projects/my-project/locations/us-east4/taskruns` as a total hypothetical)

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:260]: Yeah StepState has the same problem, but it's slightly better since there at least aren't tons of extra inherited fields. But K8s could add more any time they want, so we should spell it out defensively.

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:252]: This doc is saying it's "required" that platform implementations _support_ naming steps, not that the user's request _must_ specify one.

If there's a good clear way to describe that somewhere I'd love to clear that up for future readers.

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:307]: Yeah, this version of the doc is just to get a minimal "anything" spec out there, we can add more workspace types that platforms must support as we go.

It is _conceivable_ that a platform would only support emptyDir, though that would be a pretty serious limitation. So emptyDir should probably be the only REQUIRED option.

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:271]: Had me fooled! 😄

**Comment by ImJasonH** [comment]
Ping @vdemeester @afrittoli or anybody else who may have thoughts?

File: docs/api-spec.md
Comment (afrittoli) [reply docs/api-spec.md:50]: Right now they are grouped under "Developers" (Developers may also author these workflow configurations.) but they might be called out as workflow authors.
We do not mention these groups anywhere in the document though.  What is the purpose of having these groups here in the API spec?

File: docs/api-spec.md
Comment (afrittoli) [inline docs/api-spec.md:54]: Conformance could be associated to version numbers. 

We do not have a Tekton wide version number as such; I think it's something that we might introduce in future, at least for the pipeline + triggers couple. That would allow us to have conformance associated to an overall Tekton version - which would also be helpful for operator.

In any case having to introduce new fields as OPTIONAL matches the model of adding features behind features flags that we should follow for the API.

File: docs/api-spec.md
Comment (afrittoli) [inline docs/api-spec.md:58]: NIT: it would be nice to phrase this from an API spec POV first and then detail the K8s specific implementation. E.g. The Tekton API (or a Tekton conformant API) provides a set of API resources to manage run-to-completion workflows.
Those are implemented by Tekton as [Kubernetes Custom Resources]...

File: docs/api-spec.md
Comment (afrittoli) [inline docs/api-spec.md:66]: I'm a bit confused about the "Kubernetes implementation" terminology.
My understanding is that this documents abstracts the Tekton API with the aim to define what a "Tekton conform" API should look like.  
The reference implementation is the Tekton implementation, and it's based on k8s. There may be other implementation in future, which will not be called Tekton, and they may or may not be k8s based.
One might decide to create another k8s version of Tekton  written in Rust for instance (probably not).

File: docs/api-spec.md
Comment (afrittoli) [inline docs/api-spec.md:72]: `Cancel` relies on PATCH/REPLACE - so this excludes `cancel` from the functionality required for conformance. It's probably OK, but I wanted to highlight it anyways.

File: docs/api-spec.md
Comment (afrittoli) [inline docs/api-spec.md:89]: Why are `PipelineRuns` not included? They are part of the beta API and I see for reason for them to be excluded. 
I also have mixed feeling about having `Tasks` as OPTIONAL.
While it is possible to re-use Tasks (from the public or a private catalog) by injecting them on the fly in a `TaskRun`'s `taskSpec`, doing so would require a non-trivial amount of work for a Tekton use to move some of their workflows to such Tekton conformant platform.

File: docs/api-spec.md
Comment (afrittoli) [inline docs/api-spec.md:118]: This means that support for `taskSpec` is required for conformance, right?

File: docs/api-spec.md
Comment (afrittoli) [inline docs/api-spec.md:121]: I have mixed feelings about this. Timeouts are key to the usability of the platform... if you want to keep this optional we could mention that the platform should at least provide some timeout mechanism (which would not be workflow specific).

**Review by afrittoli — COMMENTED** [review]
Thank you for starting this!

Having everything not mentioned defaulting to OPTIONAL makes sense, it makes it harder to review though as it is not obvious what we are excluding :P I think it should be fine to iterate on this and add things incrementally as we discover them?

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:35]: 😊

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:54]: This is a really interesting concept I'd like to think on more. I do like the idea of an implementation being able to say it's "conformant as of Tekton 0.14", but that also adds a lot of complexity, for potential implementors, for conformance tests, for users just trying to build on Tekton.

How would you feel about punting on this until after this initial spec lands, and circling back?

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:58]: ❤️

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:72]: That's useful, thanks. Noted.

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:89]: PipelineRuns are excluded purely because this is the minimal set I thought we could get away with. It's _definitely_ coming, but this PR is already huge :)

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:118]: That's correct. Implementations _can_ support task references (including someday Tekton Bundles!), but they _must_ support inline-defined task specs.

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:121]: I agree! I wanted to have the initial spec be as minimally-scoped as possible, and timeout was on the bubble.

I could add a note that platforms should consider enforcing timeouts in some way, but honestly that kinda seems like their business. If they don't want to support timeout, I'm not sure we need to make them? I'm on the fence. 🤷‍♂️

File: docs/api-spec.md
Comment (ImJasonH) [reply docs/api-spec.md:61]: Yeah, I agree that's a confusing distinction. Words are hard 😅 

I'd like to avoid calling _anything_ "_the_ Tekton implementation", since the rest of this doc tries to draw the distinction that Tekton is primarily an _API_, with potentially many implementations.

I'd love to find a good term for "the implementation at https://github.com/tektoncd/pipeline", I've been referring to this as "the Kubernetes implementation", but you're right, that's not necessarily a unique identifier. Suggestions welcome, and we can also punt on finding the best term and just merge this and update it later when we find it.

**Review by vdemeester — APPROVED** [review]
Let's get this in and iterate over it :angel: 
/meow

**Pull request merged**
